05-structs.Rmd

```diff
@@ -0,0 +1,467 @@
+
+The `referenceGenome` struct contains all the variables related to the reference genome, but values cannot be defined here. The struct definition merely lays the skeleton components of the data structure, but contains no actual values.
+
+In our workflow inputs, we declare a `referenceGenome` struct variable called `refGenome` via `referenceGenome refGenome` and removed our original reference genome definitions. We can access the variables within a struct by the following syntax: `structVar.varName`, such as `refGenome.ref_name`. The [WDL syntax specs](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#struct-definition) has more information on how to define and use structs.
+
+To give values to `refGenome`, we need to modify our JSON metadata file. We define `refGenome` variable in a nested structure that corresponds to the `referenceGenome` struct. Let's take a look:
+
+```         
+{
+  "minidata_mutation_calling_v1.refGenome": {
+    "ref_fasta": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta",
+    "ref_fasta_index": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.fai",
+    "ref_dict": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.dict",
+    "ref_pac": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.pac",
+    "ref_sa": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.sa",
+    "ref_amb": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.amb",
+    "ref_ann": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.ann",
+    "ref_bwt": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.bwt",
+    "ref_name": "hg19"
+  },
+  "minidata_mutation_calling_v1.dbSNP_vcf_index": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/dbsnp_138.b37.vcf.gz.tbi",
+  ...
+}
+```
+
+Now `refGenome` has all the values it needs for our tasks.
+
+In addition, we have replaced all the reference genome inputs in `call BwaMem` with `refGenome`: we can pass information to a task via structs.
+
+Within the `BwaMem` task, we have to to refer to variables inside the struct, such as `refGenome.ref_name` (which has a value of "hg19" using this JSON metadata):
+
+```         
+task BwaMem {
+  input {
+    File input_fastq
+    referenceGenome refGenome
+  }
+  
+  String base_file_name = basename(input_fastq, ".fastq")
+  String ref_fasta_local = basename(refGenome.ref_fasta)
+
+  String read_group_id = "ID:" + base_file_name
+  String sample_name = "SM:" + base_file_name
+  String platform = "illumina"
+  String platform_info = "PL:" + platform   # Create the platform information
+
+
+  command <<<
+    set -eo pipefail
+
+    #can we iterate through a struct??
```

Comment:
Yes, but unlike an array it has to be done by key, so it's a little goofy. This should do the trick:

```
ref_files=("~{refGenome.ref_fasta}" "~{refGenome.ref_fasta_index}" "~{refGenome.ref_dict}" "~{refGenome.ref_amb}" "~{refGenome.ref_ann}" "~{refGenome.ref_bwt}" "~{refGenome.ref_pac}" "~{refGenome.ref_sa}")
for ref_file in "${ref_files[@]}"
do 
    mv "$ref_file" .
done
```

Reply:
Awesome thanks. This is something bigger that we should decide whether to use it for all of our file localization commands. Going to put this in an issue and we can talk more about it!